### PR TITLE
twister: Update path structure for out-of-tree tests

### PR DESCRIPTION
**Pull request by PerMac**
With the patch a relative path between a test and west's project top directory is always calculated instead of being relative to zephyr base. The test names and results structur change by addition of one level. This way the names/structure conventon for tests from zephyr and modules is alligned.
For external tests ".." parts are replaced with "external". This allows to maintain unified folder structure also for results of external tests. Without this patch the folder/name allignment is lost for external and module test becaming depandant on a way how twister was called.

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

### CODE REVIEW

**Comment by PerMac** [comment]
To see what the PR is about:
Copy `samples/hello_world` into some folder outside of zephyr tree.
E.g. 
```
cd ~
mkdir -p my_own_zephyr/samples
cp -r ~/zephyrproject/zephyr/samples/hello_world my_own_zephyr/samples/
```
and then
1,) `~/zephyrproject/zephyr/scripts/twister -T ~/my_own_zephyr -G`
2.) `~/zephyrproject/zephyr/scripts/twister -T ~/my_own_zephyr/samples/hello_world -G`

Without the patch one will get such structures:
1.) `~/twister-out/native_posix/samples/hello_world/sample.basic.helloworld`
2.) `~/twister-out/native_posix/sample.basic.helloworld`

with the patch both calls will result in the same structure, i.e.:
1.) `~/twister-out/native_posix/out_of_tree/my_own_zephyr/samples/hello_world/sample.basic.helloworld`
2.) `~/twister-out/native_posix/out_of_tree/my_own_zephyr/samples/hello_world/sample.basic.helloworld`

What's more, the calculated path becomes the part of the test id. Without the patch the id of OOT  tests depends on the twister call and the path used with `-T`. This decreases the traceability of test results. It also makes automatized test artifacts extracting in CI a pain in the ass, since the location of failes to store varies with the command used.

None of the above is present in the in-tree-tests. It doesn't matter then how `-T` is used since the structure always reflects the relative path to the zephyr base.

File: scripts/pylib/twister/twisterlib/testsuite.py [hunk 1]
```diff
@@ -407,13 +407,15 @@ def add_testcase(self, name, freeform=False):
...
-            # This is in ZEPHYR_BASE, so include path in name for uniqueness
-            # FIXME: We should not depend on path of test for unique names.
-            relative_ts_root = os.path.relpath(canonical_testsuite_root,
+        relative_ts_root = os.path.relpath(canonical_testsuite_root,
                                                start=canonical_zephyr_base)
-        else:
-            relative_ts_root = ""
+        # This is in ZEPHYR_BASE, so include path in name for uniqueness
+        # FIXME: We should not depend on path of test for unique names.
+        if Path(canonical_zephyr_base) not in Path(canonical_testsuite_root).parents:
```
Comment (jangalda-nsc) [inline scripts/pylib/twister/twisterlib/testsuite.py:414]: if you embed it in `is_outside_zephyr_base` code is self descriptive

File: scripts/tests/twister/test_testsuite.py [hunk 2]
```diff
@@ -93,10 +93,10 @@ def test_get_unique_exception(testsuite_root, workdir, name, exception):
             '/scripts/tests/twister/test_data/testsuites/test_b/test_b.check_1'
         ),
         (
-            os.path.join(ZEPHYR_BASE, '/scripts/tests'),
-            os.path.join(ZEPHYR_BASE, '/scripts/tests'),
+            os.path.join(ZEPHYR_BASE, 'scripts/tests'),
+            os.path.join(ZEPHYR_BASE, 'scripts/tests'),
             'test_b.check_1',
-            'test_b.check_1'
+            'scripts/tests/test_b.check_1'
```
Comment (jangalda-nsc) [inline scripts/tests/twister/test_testsuite.py:99]: I can't see test for `out_of_tree case`

**Review by jangalda-nsc — COMMENTED** [review]
This is API breaking change, so I suggest to add some parameter / config flag which turns this behavior on

File: scripts/pylib/twister/twisterlib/testsuite.py (hunk 1 above)
Comment (PerMac) [reply scripts/pylib/twister/twisterlib/testsuite.py]: Done, also reworded the comment to better reflect the workflow.

File: scripts/tests/twister/test_testsuite.py (hunk 2 above)
Comment (PerMac) [reply scripts/tests/twister/test_testsuite.py:99]: Added.

**Comment by PerMac** [comment]
@jangalda-nsc  I am not sure if this is a real "API breaking change". It is a fix for inconsistent behavior. Hence, IMO this should be the default behavior without any additional arg added to make it work.

File: scripts/pylib/twister/twisterlib/testsuite.py [hunk 3]
```diff
@@ -407,13 +407,17 @@ def add_testcase(self, name, freeform=False):
...
-            # FIXME: We should not depend on path of test for unique names.
-            relative_ts_root = os.path.relpath(canonical_testsuite_root,
+        relative_ts_root = os.path.relpath(canonical_testsuite_root,
                                                start=canonical_zephyr_base)
-        else:
-            relative_ts_root = ""
+        # Include path in name for name uniqueness.
+        # Replace all '..' with a single `out_of_tree` for out of tree tests.
+        # FIXME: We should not depend on path of test for unique names.
+        is_outside_zephyr_base  = Path(canonical_zephyr_base) not in Path(canonical_testsuite_root).parents
```
Comment (gopiotr) [inline scripts/pylib/twister/twisterlib/testsuite.py:415]: Please remove extra space before `=` sign.

File: scripts/pylib/twister/twisterlib/testsuite.py [hunk 4]
```diff
@@ -407,13 +407,17 @@ def add_testcase(self, name, freeform=False):
...
+        if is_outside_zephyr_base:
+            relative_ts_parts = Path(relative_ts_root).parts
+            if ".." in relative_ts_parts:
+                relative_ts_parts = tuple(filter(("..").__ne__, relative_ts_parts))
```
Comment (gopiotr) [inline scripts/pylib/twister/twisterlib/testsuite.py:419]: Please consider replace it by more pythonic way with list comprehension:
```python
relative_ts_parts = [part for part in relative_ts_parts if part != ".."]
```
or generator:
```python
relative_ts_parts = tuple(part for part in relative_ts_parts if part != "..")
```

**Comment by gopiotr** [comment]
I tested those changes on Windows OS and it works properly.

File: scripts/pylib/twister/twisterlib/testsuite.py (hunk 3 above)
Comment (PerMac) [reply scripts/pylib/twister/twisterlib/testsuite.py:415]: done

**Review by gopiotr — DISMISSED** [review]

**Comment by nashif** [comment]
what is out of tree here? are we talking tests in modules for example, such as mcuboot? or some folder in a random place on the disk?

**Comment by PerMac** [comment]
@nashif "out of tree" are tests locate outside of ZEPHYR_BASE directory. So both modules like mcuboot and any other user folder. The same as described in the first comment applies to e.g. a sample from zephyrproject/bootloader/mcuboot/boot/zephyr/  With main you'll get different structure in twister_out and different id of samples depending on your -T argument which breaks test trackability. It is the same test regardless of your -T usage.
I haven't thought of using mcuboot as an example but it is a valid one as well.

**Comment by nashif** [comment]
> @nashif "out of tree" are tests locate outside of ZEPHYR_BASE directory. So both modules like mcuboot and any other user folder. The same as described in the first comment applies to e.g. a sample from zephyrproject/bootloader/mcuboot/boot/zephyr/ With main you'll get different structure in twister_out and different id of samples depending on your -T argument which breaks test trackability. It is the same test regardless of your -T usage.
I haven't thought of using mcuboot as an example but it is a valid one as well.

OK, I am aware of that problematic. I do not like the `out_of_tree` terminology in this context however, what tree exactly? If you mean zephyr repo as the tree here, then it is slightly not accurate for modules, so instead of using `out_of_tree` I would call anything that is not directly managed in the zephyr workspace as `external`, this way if some reports comes in with an external test, we will know it is not managed in the zephyr project or the west workspace.

To expand the namespace and move out of the zephyr repo I would suggest to change how we calculate pathes right now and use the workspace as the top instead of the ZEPHYR_BASE, then we can have everything which is part of the zephyr project under one namespace that we can identify and locate within the project.

Someone with custom and private tests might chose to put all those tests in a seperate project maintained in the west manifest, say zephyr-extra-tests, so those will appear correctly and will be uniquely identified within that workspace.

Ideally we should get rid pf the path completely and instead improve how we identify tests, a test should be portable and we should be able to move it around and still be able to reference it and trace back to it. A semaphore test is still the same test, whether it is under zephyr/tests/kernel/semphore or /tmp/semaphore....

**Review by hakehuang — DISMISSED** [review]

**Comment by PerMac** [comment]
@nashif Thanks for comments. We are on the same page and I addressed all the issues:
- now the path is relative to west topdir. If west is not installed or project is not initialized then zephyr_base is used. This will affect the structure also for test from zephyr dir. With the patch `samples/sample.basic.helloworld` -> `zephyr/samples/hello_word/sample.basic.helloworld`
- thanks to this the structures/names for modules (e.g. bootloader) have the same pattern as ones from zephyr dir. e.g `bootloader/mcuboot/boot/zephyr/sample.bootloader.mcuboot.single_slot`
- tests outside of the project dir starts with `external` e.g. `external/my_own_zephyr/samples/hello_world/sample.basic.helloworld`

As for `Ideally we should get rid pf the path completely and instead improve how we identify tests` I also agree but won't be able to solve this with one PR. IMO the required steps are the following:
- get rid of duplicated test names. We already sent PRs solving this issue, but it might have to be repeated.
- add a workflow verifying, that if the test yaml is adressed, all the names within are unique within the whole project scope
- some order has to be made within the existing names. The naming convention is not verified which causes problem when trying to figure out where a given test belongs. We might also want to verify all new test names if they follow the convention. 
 
The last point is not a must, but I find it quite important.
Such action has to be complemented with a rework of our reporting.

File: scripts/pylib/twister/twisterlib/environment.py [hunk 5]
```diff
@@ -24,14 +24,27 @@
 if not ZEPHYR_BASE:
     sys.exit("$ZEPHYR_BASE environment variable undefined")
 
+try:
+    subproc = subprocess.run(['west', 'topdir'], check = True, stdout=subprocess.PIPE)
```
Comment (jangalda-nsc) [inline scripts/pylib/twister/twisterlib/environment.py:28]: If I remember correctly, check=True means exception is raised when exitcode !=0. That means `if subproc.returncode == 0:` is always true

**Review by jangalda-nsc — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Comment by yperess** [comment]
@PerMac @nashif believe this broke our downstream use of the `-s` flag

**Comment by yperess** [comment]
@PerMac I'm perfectly happy to be able to split up the path from the test name in something like `./scripts/twister -T path/to/test -s my.test` for a specific test or like you said `./scripts/twister -s my.test` which may run multiple tests. Thank you so much for the quick response.

**Comment by nashif** [comment]
I would rather revert this and go back to the drawing board with more input and not introduce new behaviour and worarounds to fix this.

**Comment by PerMac** [comment]
Please wait with reverting a bit. I have already a proposal to solve -s problem. Will open a PR within minutes. Maybe reverting won't be needed?

**Comment by PerMac** [comment]
@yperess @nashif  https://github.com/zephyrproject-rtos/zephyr/pull/52288 this is my idea of a solution. I think it is not a workaround in the end but expanded functionality for --test/-s.

To recap on the current situation (without wall of text):
- tests names has to be unique, so the results are not mixed/overwritten
- path is used as part of the name to guaranty a uniqueness (scenario names might not be unique)
- the above was not fulfilled for tests oustide of zephyr_base dir.
- this merged pr was an attempt to solve the issue
- it made some combinations of args not working any more (if they were based on previous behavior)
- PR #52288 is an attempt to mitigate the issue reported by @yperess by making -s more functional

**Comment by evgeniy-paltsev** [comment]
@PerMac Not fully sure how it supposed to work now - all the options I've tried with `--test` (absolute path, relative path) fails.

And I don't see changes in a documentation - so it looks like a bug for me.

**Comment by PerMac** [comment]
the PR changed cannonical name, which was used to find samples to use pathf relative to west topdir instead of a zephyr base. Hence please try `scripts/twister -s zephyr/samples/hello_world/sample.basic.helloworld`

**Comment by evgeniy-paltsev** [comment]
Ok, it seems to work, but why do we add new inconsistency here?

For west we use path without zephyr prefix:

```
west build -b nsim_hs5x_smp tests/kernel/smp -T kernel.multiprocessing.smp
```

but for twister we for some reason add it:

```
./scripts/twister -p nsim_hs5x_smp --test zephyr/tests/kernel/smp/kernel.multiprocessing.smp
```

:/

**Comment by yperess** [comment]
I agree with @evgeniy-paltsev this is very counter intuitive, when I run a test I don't want to need to know the full "name" of the test, I want to give a relative or full path to the `testcase.yaml` file + the test name. Having this ability would make twister much more like bazel (`bazel test path/to/test:test_name`)

**Comment by PerMac** [comment]
Maybe during todays testing wg meeting we can discuss how -s should work and how a canonical name should look like? I understand your points and am open for implementing what we agree on. Of course comments here and under #52288 will also be taken into an account.

**Pull request merged**